### ISSUE TITLE
Windows App Certification Unresolvable Warnings on Burn Bootstrapper

### ISSUE THREAD

**Issue by RGBradford** [wixtoolset/issues#5171]
If you validate a burn bootstrapper installer using the <a href="https://dev.windows.com/en-us/develop/app-certification-kit">Windows App Certification Kit</a>, you get these warnings in the <i>Clean reversible install test</i>:

Write appropriate Add/Remove Program values
* An optional value 'MajorVersion' is missing or invalid for program <Bundle Name> 
* An optional value 'MinorVersion' is missing or invalid for program <Bundle Name> 
* An optional value 'VersionMajor' is missing or invalid for program <Bundle Name> 
* An optional value 'VersionMinor' is missing or invalid for program <Bundle Name> 
* An optional value 'InstallLocation' is missing or invalid for program <Bundle Name> 

The requirements are outlined in section 5.5 of <a href="https://msdn.microsoft.com/en-us/windows/desktop/hh749939.aspx">Certification requirements for Windows desktop apps</a>.

**Comment by MirzaJS** [comment]
Is there a workaround for this?

**Comment by birty** [comment]
Should InstallLocation simply be set to the package cache location in order to pass the WACK certification test?  I thought about trying to get my bundled MSI package to discover the burn package's ProductCode and write the msi product location under the HKLM/CurrentVersion/Windows/Uninstall/{BURNPRODUCTCODE}/InstallLocation registry key, but really that's nonsense for anyone with multiple payloads in the bundle.
On a related note, how do I get the burn driven uninstall to be unattended? At the moment the 'unattended removal' step in WACK brings up my bundled msi's user interface and I have to click 'uninstall' to let WACK continue its checks. 
How does WACK trigger the unattended removal step? A SilentUninstallString?

**Comment by birty** [comment]
So, I forked Wix3 and modified Burn to write the package cache folder as its InstallLocation. It now passes the test referenced on this bug without warnings :) , BUT it now raises a new warning on the 'Install to correct folders' test because burn installs an executable (the cached setup.exe) to somewhere other than Program Files...  hmm.

The WACK guidelines are saying the executables should go in Program Files rather than ProgramData because it is executable code, not data.

In my case this is a per-machine install. The setup requests UAC elevation anyway, so installing the burn cache to program files will not be a problem as long as we can decide where under Program Files the burn cached packages should be stored.

My WACK results ->

Clean reversible install test
PASSED
Remove all non-shared files and folders
PASSED
Do not force an immediate reboot during installation
PASSED
Do not force an immediate reboot during uninstallation
PASSED
Write appropriate Add/Remove Program values
**PASSED  <--  PREVIOUSLY WARNED ABOUT InstallLocation**
Single user registry check
Install to the correct folders test
**WARNING <-- A NEW WARNING**
Install to Program Files
•Warning: The install to Program Files test detected the following errors:◦Program _<my app>_ fails due to install location C:\ProgramData\Package Cache\{.. bundle guid.. }\.
•Impact if not fixed:  Native apps e.g. 32 and 64 bit must be installed to the %ProgramFiles% folder by default, note that 32 bit apps installing on x64 versions of Windows must be stored under %ProgramFiles(x86)%. Storing program files in another folder can cause access and security problems for the user. 
•How to fix:  Write the files to the correct folders. See link below for more information: 
Working with Known Folders

PASSED
Write to the %WINDIR% or %SystemDrive%
PASSED

**Comment by heaths** [comment]
Probably better to allow InstallLocation to be set by a variable that a BA could set (or even a search, or by some other means) to where the main program is installed.

**Comment by birty** [comment]
Sure, but either way I think it will still fail certification due to putting an executable (the cached setup.exe) somewhere other than program files or its per-user equivalent.

**Comment by birty** [comment]
OK. Following a helpful discussion with the WIX devs I've had another attempt at this using variables.  It now passes all WACK tests except for the fact that the unattended **UN**installation step pops up the UI so I have to click Uninstall - but that's something else I need to look into.

To remove the warnings about InstallLocation I used the following technique in my BURN code update:

Defines a new internal installer variable called WixBundleInstallLocation

Burn engine attempts to auto-discover this after each successful msi installation by looking in for an InstallLocation in each MSIs own Uninstall registration area. Ensure your main product MSI package sets the ARPINSTALLLOCATION property as per the instructions here:
http://robmensching.com/blog/posts/2011/1/14/arpinstalllocation-and-how-to-set-it-with-the-wix-toolset/
If you've got multiple msis that all set an InstallLocation, the final package wins (its likely to be the main one).
In RegistrationSessionEnd, if WixBundleInstallLocation has a value, it is written as the InstallLocation in the Bundles Uninstall registration key..

WACK is happy

**Comment by birty** [comment]
I've pushed a separate commit which produces a /quiet UninstallString by specifying
`<Variable Name="UnattendedUninstall" bal:Overridable="yes" Type="numeric" Value="1" />`
in your Bundle.wxs

This allows WACK and other automated testers that rely on UninstallString to proceed without user intervention.

**Comment by barnson** [comment]
That seems really weird: `QuietUninstallString` exists for that purpose. Not sure why WACK would rely on the interactive version.

**Comment by birty** [comment]
Agreed. Maybe somebody with bigger boots than me can give it a kick...

https://connect.microsoft.com/VisualStudio/feedback/details/768967/cant-pass-the-silent-uninstallation-command-to-windows-8-sdk-app-certificate-kit-ack-s-appcert-exe-when-automating-the-validation-process

**Comment by therealhappy** [comment]
We have the same problem in our setup.

**Comment by robmen** [comment]
If this is important to you, I recommend going and voting for the Connect issue above. It captures the issue well and WACK should be updated.

**Comment by rseanhall** [comment]
https://github.com/wixtoolset/wix3/pull/331
https://github.com/wixtoolset/wix4/pull/193

Does anyone know what happened to the Connect issue? The link no longer works since they retired Connect.
